### PR TITLE
Refactor and update the token standard

### PR DESCRIPTION
**Pull request by Maximkaaa**
* Removed duplicated code between `transfer` and `tranferIncludeFee`
* Removed most 'expect's
* Renamed module erc20_transactions to icrc1_transfer
* Added some unit tests
* Added `FeeRatio` type to simplify check for correct fee ratios
* Add test cases for zero amount transactions, token amount overflows,
  self transfers etc.
* Fix a bug when self-transfer to default subaccount is allowed.
* Fix a bug when deduplication check permits a duplicated transaction if
  its `created_at_date` ts differs from IC time by some margin.
* Move claim tests from icrc1 module (they are not related to icrc).
* Changes the dispaly method to make it clear that the displayed value is
an account not a principal, and also adds the subaccount to the
displayed value. Also adds some unit tests for account.
* Removed unused error variants
* Update token standard

### CODE REVIEW

File: src/token/api/src/account.rs
Comment (blutooth) [inline src/token/api/src/account.rs:13]: why is subaccount not an opt as in the did file for the ICRC-1 spec? https://github.com/dfinity/ICRC-1/blob/main/ICRC-1.did

File: src/token/api/src/canister.rs
Comment (blutooth) [inline src/token/api/src/canister.rs:556]: so, this would just be for default accounts?

File: src/factory/src/api.rs
Comment (hagsteel) [inline src/factory/src/api.rs:75]: Can we use the actual `const` function here instead? `Principal::from_slice` I think it is

File: src/token/api/src/account.rs
Comment (hagsteel) [inline src/token/api/src/account.rs:39]: If we write the comma before the sub account it won't look so naff if there is no sub accounts
```suggestion
            write!(f, "Account({}", self.owner)?;
            for b in self.subaccount {
                write!(f, ", {b:02X}")?;
            }
```

File: src/token/api/src/canister.rs
Comment (hagsteel) [inline src/token/api/src/canister.rs:150]:
```suggestion
    /********************** BALANCES INFO ***********************/
```

File: src/token/api/src/canister/is20_transactions.rs
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:74]:
```suggestion
    // We use `updates` structure because sometimes from or to can be equal to fee_to or even to
```

File: src/token/api/src/canister/is20_transactions.rs
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:75]:
```suggestion
    // auction_account, so we must take a careful approach.
```

File: src/token/api/src/canister/is20_transactions.rs
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:84]:
```suggestion
    // If `amount + fee` overflows max `Tokens128` value, the balance cannot be larger than this
```

File: src/token/api/src/canister/is20_transactions.rs
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:177]: The comment is redundant

```suggestion
(total_supply + amount).ok_or(TxError::AmountOverflow)?;
```

File: src/token/api/src/canister/is20_transactions.rs
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:227]: Pick either, as long as it's consistent

File: src/token/api/src/error.rs
Comment (hagsteel) [inline src/token/api/src/error.rs:29]:
```suggestion
    #[error("no tokens to claim on the requested subaccount.")]
```
Also consistency: all error messages should be lower cased

File: src/token/api/src/state.rs
Comment (hagsteel) [inline src/token/api/src/state.rs:23]:
```suggestion
    // We leave this field here to not introduce a new version of the state.
    #[deprecated(note = "claims are now stored in the owner's subaccounts.")]
```

**Review by hagsteel — CHANGES_REQUESTED** [review]
Looks good. Just a few comments

File: src/token/impl/src/canister.rs
Comment (blutooth) [inline src/token/impl/src/canister.rs:159]: Copperfield

File: src/token/api/src/canister.rs
Comment (Maximkaaa) [reply src/token/api/src/canister.rs:556]: It works just like before - subaccounts can be used to specify any AccountIdentifier, but `claim` method transfers tokens to the default account of the claiming user.

File: src/token/api/src/account.rs
Comment (Maximkaaa) [reply src/token/api/src/account.rs:39]: If there's no subaccount, another brunch is executed. Adding `,` before each byte in subaccount wouldn't look nice....

File: src/token/api/src/canister/is20_transactions.rs
Comment (Maximkaaa) [reply src/token/api/src/canister/is20_transactions.rs:177]: Let me disagree on this one. Because I had to think whether I should add this check and what are the implications of not having it...

File: src/token/api/src/canister/is20_transactions.rs
Comment (Maximkaaa) [reply src/token/api/src/canister/is20_transactions.rs:227]: I just wanted to show my eloquence :)

File: src/token/api/src/canister/is20_transactions.rs
Comment (hagsteel) [reply src/token/api/src/canister/is20_transactions.rs:177]: How can `Tokens128` be more than `Tokens128::MAX`?
I don't think I know which part you are disagreeing on, could you elaborate? (I'm all for disagreement)

Comment (Maximkaaa) [reply src/token/api/src/canister/is20_transactions.rs:177]: Tokens128 cannot, but the total amount is combined of many users' balances, so if we don't check it, total balance can go over Tokens128::MAX.

**Review by blutooth — APPROVED** [review]

**Review by blutooth — APPROVED** [review]

**Comment by blutooth** [comment]
PR looks good, before we merge can we  update the tags on the AMM to a tag of this branch with a PR on the AMM to check it build and the tests pass correctly.

**Comment by Maximkaaa** [comment]
I've created a draft PR for AMM: https://github.com/infinity-swap/v1-amm/pull/224. All tests pass locally except one, but it cannot be related to IS20 update. Waiting for CI build.

**Comment by Maximkaaa** [comment]
(My AMM PR doesn't pass on CI because of the broken CI script when using branch dependency....)

**Pull request merged**
